PreProcessing/identify_kinematic_chains.m

```diff
@@ -0,0 +1,331 @@
+
+    coordi = nan(length(coords_limb_i),1);
+    for j=1:length(coords_limb_i)
+        coordi(j) = model_info.ExtFunIO.coordi.(coords_limb_i{j});
+    end
+    
+    jointi.(['leg' num2str(i) '_r']) = coordi';
+    leg_r = [leg_r coordi];
+
+    limbs.all.joints = [limbs.all.joints; joints_limb_i];
+    limbs.all.coords = [limbs.all.coords; coords_limb_i];
+    
+    [joints_limb_i, coords_limb_i] = getKinematicChain(joint_table,base_joint_l);
+
+    coordi = nan(length(coords_limb_i),1);
+    for j=1:length(coords_limb_i)
+        coordi(j) = model_info.ExtFunIO.coordi.(coords_limb_i{j});
+    end
+
+    jointi.(['leg' num2str(i) '_l']) = coordi';
+    leg_l = [leg_l coordi];
+
+    limbs.all.joints = [limbs.all.joints; joints_limb_i];
+    limbs.all.coords = [limbs.all.coords; coords_limb_i];
+end
+
+for i=1:length(base_arms)
+    base_joint_idx = find(strcmp(joint_table.joint,base_arms{i}));
+    if isempty(base_joint_idx)
+        base_joint_idx = find(strcmpi(joint_table.joint,[base_arms{i} '_r'])|strcmpi(joint_table.joint,['r_' base_arms{i}]));
+        if isempty(base_joint_idx)
+            error([base_arms{i} ' is not recognised as a joint name on the right side.']);
+        end
+    end
+    [base_joint_l, base_joint_r] = mirrorName(joint_table.joint{base_joint_idx});
+
+    [joints_limb_i, coords_limb_i] = getKinematicChain(joint_table,base_joint_r);
+
+    coordi = nan(length(coords_limb_i),1);
+    for j=1:length(coords_limb_i)
+        coordi(j) = model_info.ExtFunIO.coordi.(coords_limb_i{j});
+    end
+    
+    jointi.(['arm' num2str(i) '_r']) = coordi';
+    arm_r = [arm_r coordi];
+
+    limbs.all.joints = [limbs.all.joints; joints_limb_i];
+    limbs.all.coords = [limbs.all.coords; coords_limb_i];
+    
+    [joints_limb_i, coords_limb_i] = getKinematicChain(joint_table,base_joint_l);
+
+    coordi = nan(length(coords_limb_i),1);
+    for j=1:length(coords_limb_i)
+        coordi(j) = model_info.ExtFunIO.coordi.(coords_limb_i{j});
+    end
+    
+    jointi.(['arm' num2str(i) '_l']) = coordi';
+    arm_l = [arm_l coordi];
+
+    limbs.all.joints = [limbs.all.joints; joints_limb_i];
+    limbs.all.coords = [limbs.all.coords; coords_limb_i];
+end
+
+jointi.leg_r = leg_r';
+jointi.leg_l = leg_l';
+jointi.arm_r = arm_r';
+jointi.arm_l = arm_l';
+
+%% Find floating base
+jointi.floating_base = [];
+jointi.base_forward = [];
+jointi.base_vertical = [];
+jointi.base_lateral = [];
+
+for i=1:n_joints
+
+    idx = find(contains(joint_table.child,joint_table.parent(i)));
+    if isempty(idx)
+        floating_base = jointSet.get(joint_table.joint(i));
+        continue
+    end
+end
+
+fl_b = floating_base.getConcreteClassName();
+if strcmp(fl_b,'CustomJoint')
+    floating_base = CustomJoint.safeDownCast(floating_base);
+elseif strcmp(fl_b,'PlanarJoint')
+    floating_base = PlanarJoint.safeDownCast(floating_base);
+end
+
+sptr = floating_base.getSpatialTransform();
+
+for j=1:6
+
```

Comment:
do we need j=1:6 if it is a planar joint as well?

Reply:
Planar joint for floating base only happens in 2D models, which are not supported by OpenSimAD anyways.
Support for 2D models is included in #107.